sdk/src/azure/core/az_mqtt_connection_hfsm.c

```diff
@@ -0,0 +1,384 @@
+      }
+
+      _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, started, idle));
+      _az_RETURN_IF_FAILED(
+          az_event_policy_send_inbound_event((az_event_policy*)this_policy, event));
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result connecting(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt_connection* this_policy = (az_mqtt_connection*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_mqtt_connection/started/connecting"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+    case AZ_HFSM_EVENT_EXIT:
+      // No-op.
+      break;
+
+    case AZ_MQTT_EVENT_CONNECT_RSP:
+    {
+      az_mqtt_connack_data* data = (az_mqtt_connack_data*)event.data;
+
+      if (data->connack_reason == 0)
+      {
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, connecting, connected));
+        _az_RETURN_IF_FAILED(
+            az_event_policy_send_inbound_event((az_event_policy*)this_policy, event));
+      }
+      else
+      {
+        // TODO_L: Implement credential rotation & retry.
+        _az_RETURN_IF_FAILED(_az_hfsm_send_event(
+            (_az_hfsm*)me, (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = NULL }));
+
+        _az_RETURN_IF_FAILED(
+            az_event_policy_send_inbound_event((az_event_policy*)this_policy, event));
+      }
+
+      if (az_result_failed(_az_mqtt_connection_api_callback(this_policy, event)))
+      {
+        // Callback failed: fault the connection object.
+        _az_RETURN_IF_FAILED(_az_hfsm_send_event(
+            (_az_hfsm*)me, (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = NULL }));
+      }
+      break;
+    }
+
+    case AZ_EVENT_MQTT_CONNECTION_CLOSE_REQ:
+      _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, connecting, disconnecting));
+      _az_RETURN_IF_FAILED(_disconnect(this_policy));
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result reconnect_timeout(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  (void)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_mqtt_connection/started/reconnect_timeout"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      // TODO_L: reconnect timer (apply reconnect policy), counter not implemented.
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      // TODO_L
+      break;
+
+    case AZ_HFSM_EVENT_TIMEOUT:
+      // TODO_L
+      break;
+
+    default:
+      // TODO_L
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result connected(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  (void)me;
```

Comment:
```suggestion
```
Should this be removed since we use `me` lower down?

Reply:
Removed!